answer2_Swift.playground/Contents.swift

```diff
@@ -0,0 +1,253 @@
+        weight ／ [ln(current_time - last_accessed_time + 1) + 1]
+ 
+ Implementation of the cache should be optimized on the time complexity of get(key).
+ For example, the average time complexity of get(key) could be constant.
+ For the purpose of implementing the cache, you could assume that common data structures, such as arrays, different types of lists, and hash tables, are available.
+ At the end of the answer, give and explain the computational complexity of get(key) and put(...) in Big O notation.
+ */
+//:----
+//: ### Analyze
+/*
+ Question 2 is a variant of the LRU Cache.
+ 
+ Step0: Complete LRU Cache
+ */
+//:----
+//: ## LRU Cache
+//: ### Analyze
+/*
+ LRU Cache retain a sorted and finite amount of data. The data is sorted in the order by the time it been accessed.
+ When the capacity is full, the least recently used object of data will be delete.
+ 
+ Step0: A payload protocol to contain key and value for node.
+ Step1: CachePayload for node to store.
+ Step2: Node contain payload, previous link and next link.
+ Step3: Doubly Linked List insert, remove, MoveTo
+ */
+//:----
+protocol Payload {
+    associatedtype Key
+    associatedtype Value
+    
+    var key: Key { get set }
+    var value: Value { get set }
+}
+
+struct CachePayload<T: Hashable, U>: Payload {
+    
+    var key: T
+    var value: U
+    
+    init(key: Key, value: Value) {
+        self.key = key
+        self.value = value
+    }
+}
+
+class Node<T: Payload> {
+    
+    var payload: T
+    var previous: Node<T>?
+    var next: Node<T>?
+    
+    init(value: T) {
+        self.payload = value
+    }
+}
+
+class DoublyLinkedList<T: Payload> {
+    
+    // The head starting at index 0.
+    var head: Node<T>?
+    var tail: Node<T>?
+    var isEmpty: Bool {
+        return head == nil && tail == nil
+    }
+    
+    public var count: Int = 0
+    
+    public func insertHead(node: Node<T>) {
+        guard !isEmpty else {
+            head = node
+            tail = node
+            count += 1
+            return
+        }
+        
+        var currNode = head
+        
+        if currNode === head {
+            if head === tail {
+                node.next = head
+                head?.previous = node
+                head = node
+            } else {
+                node.next = head
+                head = node
+            }
+            
+            count += 1
+            return
+        }
+        
+        node.previous = currNode
+        node.next = currNode?.next
+        currNode?.next?.previous = node
+        currNode?.next = node
+        
+        count += 1
+        return
+    }
+    
+    func removeTail() {
+        guard !isEmpty else {
+            return
+        }
+        
+        if head === tail {
+            head = nil
+            tail = nil
+            count -= 1
+            return
+        }
+        
+        tail?.previous?.next = nil
+        tail = tail?.previous
+        
+        count -= 1
+        return
```

Comment:
通常會建議把移除的節點回傳出來

```suggestion
    @discardableResult
    func removeTail() -> Node<T>? {
        guard !isEmpty else {
            return nil
        }
        let node = head
        if head === tail {
            head = nil
            tail = nil
            count -= 1
            return node
        }
        
        tail?.previous?.next = nil
        tail = tail?.previous
        
        count -= 1
        return node
```

ps. `@discardableResult` 是讓你可以把這個函式當成沒有回傳值的函式